File: automator/src/main/java/com/stardust/automator/filter/ListFilter.java
package com.stardust.automator.filter;

import com.stardust.automator.UiObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Stardust on 2017/3/9.
 */

public interface ListFilter {

    List<UiObject> filter(List<UiObject> nodes);

    abstract class Default implements Filter, ListFilter {

        @Override
        public List<UiObject> filter(List<UiObject> nodes) {
            List<UiObject> list = new ArrayList<>();
            for (UiObject node : nodes) {
                list.addAll(filter(node));
            }
            return list;
        }
    }

}

File: automator/src/main/java/com/stardust/automator/filter/ListFilter.kt
package com.stardust.automator.filter

import com.stardust.automator.UiObject

import java.util.ArrayList

/**
 * Created by Stardust on 2017/3/9.
 */

abstract class ListFilter {

    var maxCount: Int = Integer.MAX_VALUE

    abstract fun filter(nodes: List<UiObject>): List<UiObject>

    abstract class Default : Filter, ListFilter() {

        override fun filter(nodes: List<UiObject>): List<UiObject> {
            val list = ArrayList<UiObject>()
            for (node in nodes) {
                list.addAll(filter(node))
                if (list.size >= maxCount) {
                    break
                }
            }
            return list
        }
    }

}
